### ISSUE TITLE
Fails to play any episodes/programs on Dplay.se

### ISSUE THREAD

**Issue by orfeous1** [retrospect-addon/plugin.video.retrospect#1364]
Not able to play any program/episode on Dplay.se

[retrospect.log](https://github.com/retrospect-addon/plugin.video.retrospect/files/4553533/retrospect.log)

**Comment by basrieter** [comment]
You are trying to play a premium video. Indicated by the little golden "a" after the title. Other titles play just fine. You can filter the premium items by setting the corresponding filter setting the Retrospect add-on settings.

**Comment by orfeous1** [comment]
No, that is incorrect!
Ive tried shows/episodes that works to play on Dplay.se website.

**Comment by basrieter** [comment]
![image](https://user-images.githubusercontent.com/1310443/80638881-735a9b00-8a61-11ea-8978-ac272bd8ab9b.png)

The one from the log was a premium one. So please provide examples of those shows that should play according to you.

**Comment by basrieter** [comment]
I don't support logging in, so Registered and Premium won't work for now. Do you have a login for dplay?

**Comment by orfeous1** [comment]
It is still free to watch, but they recently changed that you are required to login with a free account to watch. No premium/pay as i told you several times!

**Comment by orfeous1** [comment]
![Screenshot_20200504-211431_Chrome](https://user-images.githubusercontent.com/22922189/81004584-e80c4b80-8e4c-11ea-83b8-7b044ddd370d.jpg)

**Comment by basrieter** [comment]
But they do require a login. And that is an issue, as their login is absolutely not straight forward.....

**Comment by orfeous1** [comment]
You can use my free account if you want and if it helps?

**Comment by basrieter** [comment]
That is not the problem. The hashing algorithm seems to be none standard. Need to put in some time.

**Comment by basrieter** [comment]
Got the algorithm figured out. So that is progress.

**Comment by orfeous1** [comment]
Sounds really Good! :)
I think tv4play.se also have this login requirement in their site. But it seems to work without in retrospect. Do they work different as in login requirement stuff?

I wish i learned Python so i could help developing/testing. I only know C++

**Comment by basrieter** [comment]
I think I nailed it. Please install https://github.com/retrospect-addon/plugin.video.retrospect/archive/454d538e9ea36e4023d2dc01bfd32f121454f280.zip using the "install from zip option". You should restart Kodi, run Retrospect once and then go to the _channel settings_ of Dplay. A username/password options should be there:

![image](https://user-images.githubusercontent.com/1310443/81511169-93b70f00-9317-11ea-9a5e-44288a946629.png)

This features makes use of the Retrospect Vault to securely store your password.

**Comment by orfeous1** [comment]
Failed to install with this error message.
"requires xbmc.python version 2.25.0 which is not available"

**Comment by basrieter** [comment]
It's Leia only for now (until it's done and then it will merge into the Matrix branch).

Please install https://github.com/retrospect-addon/plugin.video.retrospect/archive/aeda2c09cf34e97bd456e53bd4a9a1211cd99c5e.zip (fixed a minor thing)

**Comment by orfeous1** [comment]
Okey, i will have to wait then for Matrix release!
I tried to install your fixed fix and faild to with same error.

**Comment by basrieter** [comment]
Well, I really need you to verify this. I only tested it here and that does make it ready for the public. Anyway you can run a Kodi Leia in Portable mode?

**Comment by orfeous1** [comment]
Yeah, did a Leia portable install, installed retrospect from repo. Installed the fix from zip.
Success.
Configured retrospect vault PIN.
Configured Dplay channel with my username and password.
Tried to play an episode of "Wahlgrens värld" but it failed.

Safe to attach log or does it include sensitive information according to username and stuff?

**Comment by basrieter** [comment]
Great. I added better `premium` detection and also enabled the login for Kanal 5,7 and 11. It does require you to set the username/password one more time. 

https://github.com/retrospect-addon/plugin.video.retrospect/archive/22c08381166d780b5d3f3efc9bdd31b4e4bfdc93.zip

Let me know if this works.

**Comment by basrieter** [comment]
@orfeous1 @sopor could you please try the last link? You will need to set new credentials. If it works, I will merge.

**Comment by Snaah** [comment]
@basrieter This is great! I emptied the credential and it works here.

Only thing I'm thinking is why have separate entries for the other DPlay channels? Isn't that unnecessary? One login to rule them all! :)

**Comment by basrieter** [comment]
They are all the same for Dplay/Kanal5,7,9,11. Set one, and all of them should get it.

**Comment by Sopor** [comment]
Seems to work ...

**Comment by Snaah** [comment]
> They are all the same for Dplay/Kanal5,7,9,11. Set one, and all of them should get it.

Yes, I saw that. Just saying the settings for the other channels aren't necessary. I'd suggest having the setting for DPlay say it sets the login for the other channels or the individual channel settings refer to the DPlay setting. Confusing for the user that isn't advanced.

**Comment by basrieter** [comment]
> > They are all the same for Dplay/Kanal5,7,9,11. Set one, and all of them should get it.
> 
> Yes, I saw that. Just saying the settings for the other channels aren't necessary. I'd suggest having the setting for DPlay say it sets the login for the other channels or the individual channel settings refer to the DPlay setting. Confusing for the not so advanced user with a setting for each channel.

That's not how it works (and was designed). The settings are generated automatically and can be set from each channel that uses them.

**Comment by basrieter** [comment]
I also updated the official Leia and Matrix branches to include this change. This makes installing it easier for now. Hope to release a new version of Retrospect soon. 

Leia: https://github.com/retrospect-addon/plugin.video.retrospect/archive/kodi-leia.zip
Matrix: https://github.com/retrospect-addon/plugin.video.retrospect/archive/kodi-matrix.zip

**Comment by Snaah** [comment]
> > > They are all the same for Dplay/Kanal5,7,9,11. Set one, and all of them should get it.
> > 
> > 
> > Yes, I saw that. Just saying the settings for the other channels aren't necessary. I'd suggest having the setting for DPlay say it sets the login for the other channels or the individual channel settings refer to the DPlay setting. Confusing for the not so advanced user with a setting for each channel.
> 
> That's not how it works (and was designed). The settings are generated automatically and can be set from each channel that uses them.

Setting login and password for any of those channels changes the login for the other ones, right? Then there's no need to have more than one login setting - DPlay. 

Instead remove the Kanal channels from channels settings and have info for the DPlay channel that it sets the login for Kanal 5, Kanal 9 and Kanal 11 too
OR
keep the Kanal channels and have channel settings for those say the login is set with DPlay.

Just my 50 cents to make Retrospect more user friendly.

**Comment by basrieter** [comment]
Next time...for now I am done with DPlay. It really took me a lot of time (and late nights) to work out their authentication.

**Comment by Snaah** [comment]
Okay. When you do a login for TV4-gruppen or Viafree it will maybe happen then.

Thanks for your work!

**Comment by jnsson** [comment]
I get "Incorrect Retrospect PIN specified" when I try to log in to DPlay in Retrospect settings. I know it's the right password and email adress. So is it something else I should put in the password instead of the password I have to DPlay?

**Comment by orfeous1** [comment]
You need to set a PIN in the settings
 I had the same issue until i figure it out.
PIN and password för dplay are separate things

**Comment by jnsson** [comment]
Right. So a PIN in Kodi settings only? Because in Restrospect it says, email and password. So should I sett a PIN in Kodi or where is it I should set a PIN?

**Comment by orfeous1** [comment]
"Inställningar", "Allmänt", "Lösenordsskydd", "Ändra koden"

**Comment by jnsson** [comment]
Tack!! (Thanks!!) Do I need to enter the PIN everytime I start a video from DPlay in Retrospect? Because I was asked for a PIN when I started a video now.

**Comment by orfeous1** [comment]
Only once :)

**Comment by jnsson** [comment]
Ok. So when I start Kodi or rather a DPlay video next week, I shouldn't be asked for a PIN, correct?

**Comment by orfeous1** [comment]
I dont think so. I dont remember. I havent put it in anymore. Put it in if its ask for it :)

**Comment by jnsson** [comment]
Yes, A bit annoying only :). But then I know. Thanks for the help!

**Comment by Sopor** [comment]
The PIN is there to encrypt the password so it is not saved in plain text.

**Comment by basrieter** [comment]
Regarding the pin: please read https://www.rieter.net/content/2017/02/07/storing-passwords-in-the-retrospect-vault/

**Comment by Sopor** [comment]
I still think it is overkill to encrypt the password. It will only create annoyance and frustration to have this PIN. Maybe 1 per mille have use for it, so it would have been nice if it was possible to disable the PIN.

**Comment by basrieter** [comment]
It won't be removed. Storing passwords in plain text is just dead wrong. So please stop asking for it.

**Comment by Espancor** [comment]
I don´t get it! Do I need to have a Dplay account to watch Kanal 5 programms? They ask me about a PIN.

**Comment by orfeous1** [comment]
Yes, dplay account is required. They recently changed it to be required. Its free.
You can only see free episodes though. They changed that too. Thumbs down for the whole dplay situation.

PIN needs to be changed in retrospect settings.

Dplay account can be entered in specifik dplay channel settings in retrospect

**Comment by Espancor** [comment]
Ok, It´s sucks! @orfeous1 Thanks anyway!

**Comment by Espancor** [comment]
just a question! Last week I took a 2 weeks free of Dplay, I was force to put in a Credit card on it and just wondering if when the 2 weeks expire if I could use the same account to wathc the free programms or do I have to make a new account. Can I take away my credit cards from my account if I use just a Dplay account or Am I force to use it.
Tack!!

**Comment by Snaah** [comment]
It's of course impossible for us to answer, only Dplay knows for sure. My guess, though, is that you can continue using your account. Just make sure you cancel the subscription before the two weeks end unless you want to continue paying.
